src/lib/crypto/kimchi_backend/gadgets/ec_group.ml

```diff
@@ -0,0 +1,3965 @@
+ *     Pre-computing doubles: 767 (for 256-bit curve; one-time cost per circuit)
+ *)
+let scalar_mul (type f) (module Circuit : Snark_intf.Run with type field = f)
+    (external_checks : f Foreign_field.External_checks.t)
+    (curve : f Curve_params.InCircuit.t) ?(doubles : f Affine.t array option)
+    (scalar : Circuit.Boolean.var list) (point : f Affine.t) : f Affine.t =
+  (* Double-and-add algorithm
+   *   Only used for signature verification, so simple algorithm suffices.
+   *
+   *     A = O; B = P
+   *     for i in 0..z
+   *         if si == 1
+   *             A = group_add(A, B)
+   *         B = group_double(B)
+   *     return A
+   *
+   *   Optimization:
+   *
+   *     To avoid expensive in-circuit conditional checks for point at infinity,
+   *     we employ a randomized strategy that avoids adding the identity element
+   *     or the same point to itself.  The strategy works as follows.
+   *
+   *     Since the prover knows the the points that it wewill add and double during
+   *     scaling, the prover could select an initial accumulator point I such that
+   *     the double-and-add algorithm never adds the identity element, same point
+   *     or negated point to itself whilst scaling.
+   *
+   *     The algorithm above is modified to initialize the accumulator to I and
+   *     then (group) subtract I after scaling to compute the final result point.
+   *
+   *       A = I; B = P
+   *       for i in 0..z
+   *           if si == 1
+   *               A = group_add(A, B)
+   *           B = group_double(B)
+   *       return A + -I
+   *
+   *     The prover MUST additionally constrain that
+   *       1) point I is on the curve
+   *       2) I' = -I
+   *
+   *   Simplification:
+   *
+   *     Uniformly and randomly select initial accumulator point I, instead of using
+   *     the complicated deterministic process.
+   *
+   *     For a z-bit scalar, there are z unique B points.  Each point also has its
+   *     negative, which we cannot add to itself.  Therefore, in total there are
+   *     2z points that we do not want to select as our initial point nor compute
+   *     as an intermediate A point during scaling.  The probability we select or
+   *     compute one of these points is approx 2z^2/n, where n is the order of the
+   *     elliptic curve group.
+   *
+   *     The probability of selecting a bad point is negligible for our applications
+   *     where z is very small (e.g. 256) and n is very large (e.g. 2^256).  Thus,
+   *     we can simply randomly select the initial accumulator I and the
+   *     double-and-add algorithm will succeed with overwhelming probability.
+   *)
+  let acc, _base =
+    List.foldi scalar ~init:(curve.ia.acc, point) (* (acc, base) *)
+      ~f:(fun i (acc, base) bit ->
+        (* Add: sum = acc + base *)
+        let sum = add (module Circuit) external_checks curve acc base in
+        (* Bounds 1:
+         *   Left input is previous result, so already checked.
```

Comment:
What about the very first left input, meaning the first value of `acc`? Is it the random value that you told me about, that you subtract at the end to bypass the point at infinity representation? If so, I guess that you know that acc must be a correct point, no? Meaning that you don't need bound for it

Reply:
Yes, it's either a public circuit constant or constrained with `check_ia` gadget

Reply:
Most likely each curve can just have it's own trustlessly computed public circuit constant.